distributed/shuffle/shuffle_extension.py

```diff
@@ -329,3 +429,151 @@ def _get_shuffle(self, shuffle_id: ShuffleId) -> Shuffle:
             raise ValueError(
                 f"Shuffle {shuffle_id!r} is not registered on worker {self.worker.address}"
             ) from None
+
+    async def close(self):
+        self.executor.shutdown()
+        MultiFile.queue = None
+
+
+class ShuffleSchedulerExtension:
+    """
+    Shuffle extension for the scheduler
+
+    Today this mostly just collects heartbeat messages for the dashboard,
+    but in the future it may be responsible for more
+
+    See Also
+    --------
+    ShuffleWorkerExtension
+    """
+
+    def __init__(self, scheduler):
+        self.scheduler = scheduler
+        self.shuffles = defaultdict(lambda: defaultdict(dict))
+
+    def heartbeat(self, ws, data):
+        for shuffle_id, d in data.items():
+            self.shuffles[shuffle_id][ws.address].update(d)
+
+
+def split_by_worker(
+    df: pd.DataFrame, column: str, npartitions: int, workers: list[str]
+) -> dict:
+    """
+    Split data into many arrow batches, partitioned by destination worker
+    """
+    import numpy as np
+    import pandas as pd
+    import pyarrow as pa
+
+    grouper = (len(workers) * df[column] // npartitions).astype(df[column].dtype).values
+
+    t = pa.Table.from_pandas(df)
+    del df
+    t = t.add_column(len(t.columns), "_worker", [grouper])
+    t = t.sort_by("_worker")
+
+    worker = np.asarray(t.select(["_worker"]))[0]
+    t = t.drop(["_worker"])
+    splits = np.where(worker[1:] != worker[:-1])[0] + 1
+    splits = np.concatenate([[0], splits])
+
+    shards = [
+        t.slice(offset=a, length=b - a) for a, b in toolz.sliding_window(2, splits)
+    ]
```

Comment:
@mrocklin FYI: https://issues.apache.org/jira/browse/ARROW-14607

Not sure this is actually useful (don't see an API to get the groups out of a [`TableGroupBy`](https://arrow.apache.org/docs/python/generated/pyarrow.TableGroupBy.html#pyarrow.TableGroupBy))? Based on my experience trying to do this with pyarrow compute, I also wouldn't be surprised if performance was better in NumPy anyway.

Reply:
Yeah, they only do aggregates for now.  It's ok though, this is no longer part of the critical path.  I've mostly moved on from this problem.  It would be great to speed things up, but I don't think that it will have a significant impact.